atmcfg/atmcfg.go

```diff
@@ -44,37 +56,120 @@ func setDisabledByReplicaSetName(out *opsmngr.AutomationConfig, name string, dis
 		return s.Name == name
 	})
 	if found {
 		s := out.Sharding[i]
 		// disable shards
 		for _, rs := range s.Shards {
-			setDisabledByReplicaSetName(out, rs.ID, disabled)
+			setDisabledByReplicaSetNameAndProcesses(out, rs.ID, processesMap, disabled)
 		}
 		// disable config rs
-		setDisabledByReplicaSetName(out, s.ConfigServerReplica, disabled)
+		setDisabledByReplicaSetNameAndProcesses(out, s.ConfigServerReplica, processesMap, disabled)
 		// disable mongos
 		for i := range out.Processes {
 			if out.Processes[i].Cluster == name {
-				out.Processes[i].Disabled = disabled
+				setDisable(out.Processes[i], processesMap, disabled)
 			}
 		}
 	}
 }
 
+func setDisable(process *opsmngr.Process, processesMap map[string]bool, disabled bool) {
+	if len(processesMap) == 0 {
+		process.Disabled = disabled
+		return
+	}
+
+	key := fmt.Sprintf("%s:%d", process.Hostname, process.Args26.NET.Port)
+	if _, ok := processesMap[key]; ok {
+		process.Disabled = disabled
+		processesMap[key] = true
+	}
+}
+
 // Shutdown disables all processes of the given cluster name.
-func Shutdown(out *opsmngr.AutomationConfig, name string) {
-	setDisabledByClusterName(out, name, true)
+func Shutdown(out *opsmngr.AutomationConfig, clusterName string) {
+	setDisabledByClusterName(out, clusterName, true)
+}
+
+// ShutdownProcessesByClusterName disables the entire cluster or its processes. Processes are provided in the format {"hostname:port","hostname2:port2"}.
+func ShutdownProcessesByClusterName(out *opsmngr.AutomationConfig, clusterName string, processes []string) error {
+	if len(processes) == 0 {
+		Shutdown(out, clusterName)
+		return nil
+	}
+	return shutdownProcesses(out, clusterName, processes)
+}
+
+// shutdownProcesses disables processes. Processes are provided in the format {"hostname:port","hostname2:port2"}.
+func shutdownProcesses(out *opsmngr.AutomationConfig, clusterName string, processes []string) error {
+	processesMap := map[string]bool{}
+	for _, hostnameAndPort := range processes {
+		processesMap[hostnameAndPort] = false
+	}
+
+	setDisableByNameAndProcesses(out, clusterName, processesMap, true)
+
+	return newProcessNotFoundError(clusterName, processesMap)
+}
+
+func setDisableByNameAndProcesses(out *opsmngr.AutomationConfig, clusterName string, processesMap map[string]bool, disabled bool) {
+	newDeploymentAuthMechanisms(out)
+	setDisabledByReplicaSetNameAndProcesses(out, clusterName, processesMap, disabled)
+	setDisabledByShardNameAndProcesses(out, clusterName, processesMap, disabled)
+}
+
+// newProcessNotFoundError returns an error if a process was not found.
+func newProcessNotFoundError(clusterName string, processesMap map[string]bool) error {
+	var processesNotFound []string
+	for k, v := range processesMap {
+		if !v {
+			processesNotFound = append(processesNotFound, k)
+		}
+	}
+
+	if len(processesNotFound) > 0 {
+		return &ProcessNotFoundError{Err: fmt.Errorf(`process not found in "%v": %v`, clusterName, processesNotFound)}
```

Comment:
[nit] almost there but it can be simplified given you don't need extra fields in the struct, check in this file for `ErrUnsupportedMechanism` and how it's used